Actor events for the verified registry

aarshkshah1992 opened the pull request:
Subsumes #1320 .

Based on discussions at #1320, https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0049.md AND https://github.com/filecoin-project/FIPs/discussions/754.

Implements events with minimal payload wherein the client has just enough information to determine whether to query state or not based on the payload of the recieved event.

anorth commented:
I've made a new integration branch `integration/actor-events`. Please rebase this PR against that branch, and we'll do subsequent code reviews against it too.

anorth commented on actors/verifreg/src/emit.rs:
```diff
@@ -0,0 +1,86 @@
+impl WithAllocation for EventBuilder {
+    fn with_allocation(self, id: AllocationID, alloc: &Allocation) -> EventBuilder {
+        self.field_indexed("id", &id)
+            .field_indexed("client", &alloc.client)
+            .field_indexed("provider", &alloc.provider)
+            .field_indexed("data-cid", &alloc.data)
```
Based on https://github.com/filecoin-project/FIPs/discussions/754#discussioncomment-7204690, I don't think we should add any fields beyond ID. Which also means you can remove this whole method. Ditto for claims.

aarshkshah1992 replied:
@anorth Is an Allocation Id unique across ALL allocations ? Same question for claims.

aarshkshah1992 replied:
@anorth Didn't realise that an Allocation/Claim ID is unique across all allocation/claims. If that is the case, why do we need more fields for looking up a claim/allocation here in `get_allocation` AND `get_claim` here ?

https://github.com/filecoin-project/builtin-actors/blob/1e50065123c05fd4881cc1898d59666c3bb97233/actors/verifreg/src/state.rs#L240

anorth replied:
Yes they are unique, but they are not indexed directly. The only lookup available is via (client, allocation) or (provider, claim).

anorth commented on runtime/src/util/events.rs:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    /// Initialise the "type" of the event i.e. Actor event type.
+    pub fn event_type(self, _type: &str) -> Self {
```
Nit: `event_type` redundantly says "event" - this context will almost always be obvious because it'll be a chained method call from `EventBuilder::new().event_type()`. So I suggest just `type` (or `typ` if that's a keyword).

aarshkshah1992 replied:
This is done.

anorth commented on runtime/src/util/events.rs:
```diff
@@ -0,0 +1,127 @@
+
+    /// Initialise the "type" of the event i.e. Actor event type.
+    pub fn event_type(self, _type: &str) -> Self {
+        self.push_entry(EVENT_TYPE_KEY, _type, Flags::FLAG_INDEXED_KEY)
```
I think this should be indexed key and value. I expect that looking up events of a certain type would be very common.

aarshkshah1992 replied:
Done.

anorth commented on runtime/src/util/events.rs:
```diff
@@ -0,0 +1,127 @@
+
+    #[test]
+    fn event_type() {
+        let e = EventBuilder::new().event_type("l1").event_type("l2").build().unwrap();
```
Two types on one event would be pretty uncommon, so it's odd as a test case.

aarshkshah1992 replied:
Done with a better test.

anorth reviewed (commented):
This is a cursory review mainly of design aspects. I'll review all the actual events emitted in detail once those questions settle.

aarshkshah1992 commented:
@anorth Have addressed your review and rebased this against the events integration branch. Please take a look.

anorth commented on actors/verifreg/src/lib.rs:
```diff
@@ -328,12 +335,18 @@ impl Actor {
+
+                    emit::allocation_removed(rt, *id)?;
+
                     // Unwrapping here as both paths to here should ensure the allocation exists.
```
Please move this comment up  next to the unwrap().

aarshkshah1992 replied:
Done.

anorth commented on actors/verifreg/src/lib.rs:
```diff
@@ -542,11 +558,12 @@ impl Actor {
 
                     let new_claim = Claim { term_max: term.term_max, ..*claim };
-                    st_claims.put(term.provider, term.claim_id, new_claim).context_code(
+                    st_claims.put(term.provider, term.claim_id, new_claim.clone()).context_code(
```
Is this clone redundant now?

aarshkshah1992 replied:
Yes it is, removed.

anorth reviewed (approved):
This looks good.

I have a design question for you both @aarshkshah1992 @Stebalien. Should we restrict emitting events inside state transactions (in the runtime lib)?

An event is a side-effect, but the state transaction might roll back (without the message necessarily aborting). There's no re-entrancy risk that makes this so dangerous with sending messages, but still potential for error emitting an event for a state transition that didn't happen.

If yes, we should enforce this in the runtime, and plumb some of the event field values out of the transactions from which they're currently emitted and adopt a patter of emitting events at the end of method execution. We can do this in a follow-up.

Stebalien commented:
> An event is a side-effect, but the state transaction might roll back (without the message necessarily aborting). There's no re-entrancy risk that makes this so dangerous with sending messages, but still potential for error emitting an event for a state transition that didn't happen.

If an actor reverts (exits with a non-zero exit code), all emitted events will also disappear. However, we _do_ have to be careful about cases where an actor might:

1. Encounter some error while updating state.
2. Decide to skip the state update but continue without reverting.

But I'm not sure if refusing to emit events during transactions is really the right solution here (or if this is even a "real" issue in practice).

aarshkshah1992 commented:
@anorth Have addressed your review. 

Based on what @Stebalien is saying here, I think we need to be careful in the code/review that we don't (encounter error -> NOT revert -> but emit event). Looks like transaction rollbacks already rollback events as well. 

If there are any action items coming out of this, we can do it in a follow-up PR.
